iroh-net/src/key/encryption.rs

```diff
@@ -0,0 +1,153 @@
+//! The private and public keys of a node.
+
+use std::fmt::Debug;
+
+use anyhow::{anyhow, ensure, Result};
+
+pub(crate) const NONCE_LEN: usize = 24;
+
+fn public_ed_box(key: &ed25519_dalek::VerifyingKey) -> crypto_box::PublicKey {
+    crypto_box::PublicKey::from(key.to_montgomery())
+}
+
+fn secret_ed_box(key: &ed25519_dalek::SigningKey) -> crypto_box::SecretKey {
+    crypto_box::SecretKey::from(key.to_scalar())
+}
+
+/// Shared Secret.
+pub struct SharedSecret(crypto_box::ChaChaBox);
+
+impl Debug for SharedSecret {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "SharedSecret(crypto_box::ChaChaBox)")
+    }
+}
+
+impl SharedSecret {
+    fn new(this: &crypto_box::SecretKey, other: &crypto_box::PublicKey) -> Self {
+        SharedSecret(crypto_box::ChaChaBox::new(other, this))
+    }
+
+    /// Seals the provided cleartext.
+    pub fn seal(&self, cleartext: &[u8]) -> Vec<u8> {
+        use crypto_box::aead::{Aead, AeadCore, OsRng};
+
+        let nonce = crypto_box::ChaChaBox::generate_nonce(&mut OsRng);
+        let ciphertext = self
+            .0
+            .encrypt(&nonce, cleartext)
+            .expect("encryption failed");
+
+        let mut res = nonce.to_vec();
+        res.extend(ciphertext);
+        res
+    }
+
+    /// Opens the ciphertext, which must have been created using `Self::seal`, and returns the cleartext.
+    pub fn open(&self, seal: &[u8]) -> Result<Vec<u8>> {
+        use crypto_box::aead::Aead;
+        ensure!(seal.len() > NONCE_LEN, "too short");
+
+        let (nonce, ciphertext) = seal.split_at(NONCE_LEN);
+        let nonce: [u8; NONCE_LEN] = nonce.try_into().unwrap();
+        let plaintext = self
+            .0
+            .decrypt(&nonce.into(), ciphertext)
+            .map_err(|e| anyhow!("decryption failed: {:?}", e))?;
+
+        Ok(plaintext)
+    }
+}
+
+impl crate::key::Keypair {
```

Comment:
maybe `super::Keypair`?

Reply:
super doesn‘t work, that is the top level already